general/pionerv1_price.js

```diff
@@ -0,0 +1,177 @@
+                const prices = await this.fetchPrices(asset1, asset2);
+                const isMarketOpen = this.isValidPriceData(prices, /* appropriate config */);
+
+                return {
+                    asset1: this.convertToBytes32(asset1),
+                    asset2: this.convertToBytes32(asset2),
+                    pairBid: this.scaleUp(prices.pairBid).toString(),
+                    pairAsk: this.scaleUp(prices.pairAsk).toString(),
+                    confidence: this.scaleUp(prices.confidence).toString()
+
+                };
+        }
+    },
+
+    signParams: function (request, result) {
+        const signTime = result.oldestTimestamp ? result.oldestTimestamp : Math.floor(Date.now() / 1000);
+
+        switch (request.method) {
+            case 'price':
+                return [
+                    { name:'asset1' ,type: 'bytes32', value: result.asset1 },
+                    { name:'asset2' ,type: 'bytes32', value: result.asset2 },
+                    { name:'pairBid' ,type: 'uint256', value: result.pairBid },
+                    { name:'pairAsk' ,type: 'uint256', value: result.pairAsk },
+                    { name:'confidence' ,type: 'uint256', value: result.confidence },
+                    { name:'signTime' ,type: 'uint256', value: signTime.toString() }
+                ];
+        }
+    },
+
+fetchPrices: async function (asset1, asset2) {
+    const [result1, result2] = await Promise.all([
+        this.fetchAssetPrices(asset1),
+        this.fetchAssetPrices(asset2)
+    ]);
+
+    const adjustedPrices1 = asset1 === 'hardusd' ? { avgBid: 1, avgAsk: 1 } : this.calculateAveragePrices(result1.prices);
+    const adjustedPrices2 = asset2 === 'hardusd' ? { avgBid: 1, avgAsk: 1 } : this.calculateAveragePrices(result2.prices);
+    const asset1Confidence = asset1 === 'hardusd' ? 0 : this.calculateConfidence(result1.prices);
+    const asset2Confidence = asset2 === 'hardusd' ? 0 : this.calculateConfidence(result2.prices);
+    const highestConfidence = Math.max(asset1Confidence, asset2Confidence);
+    const oldestTimestamp = Math.min(result1.oldestTimestamp, result2.oldestTimestamp);
+
+    return {
+        pairBid: adjustedPrices1.avgBid / adjustedPrices2.avgBid,
+        pairAsk: adjustedPrices1.avgAsk / adjustedPrices2.avgAsk,
+        confidence: Math.max(1 - highestConfidence / 100, 0),
+        oldestTimestamp
+    };
+},
+
+    fetchAssetPrices: async function (asset) {
+        if (asset === 'hardusd') return { prices: [{ bid: 1, ask: 1, timestamp: null }], oldestTimestamp: null };
+        const [assetType, assetSymbol] = asset.split('.');
+        const apiConfigs = this.loadApiConfigsForType(assetType);
+        let oldestTimestamp = Infinity;
+        const prices = [];
+
+        for (const config of apiConfigs) {
+            const formattedSymbol = this.formatSymbolForAPI(assetType, assetSymbol);
+            const priceData = await this.fetchPriceFromAPI(config, formattedSymbol);
```

Comment:
This function will run on all nodes of the TSS subnet. The API response should be consistent on all nodes. Otherwise, each node will sign different data and the final signature cannot be generated.

If the response may change during a request, a solution like this app should be implemented to allow a small margin in the prices:
https://github.com/muon-protocol/muon-apps/blob/master/general/v3_oracle.js